ROFL revamp for marketplace and ROFL node (#1261)
oasisprotocol/docs: docs/node/run-your-node/prerequisites/set-up-trusted-execution-environment-tee.md

:::info

> In case the ParaTime or ROFL you want to run does not require the use of a TEE

Reviewer: ROFL always requires the use of a TEE, so we should remove this.

(e.g. Intel SGX/TDX), you can skip setting up a TEE.

:::

Most Oasis ParaTimes and ROFLs are configured to run in the TEE. There are two

> generations of Intel TEEs currently in use:

Reviewer: ```suggestion
kinds of Intel TEEs currently in use:
```

- the [Intel SGX] is required by the ParaTimes and [ROFL SGX][rofl-types] apps.

> - the [Intel TDX] is required by the [ROFL TDX raw and containers][rofl-types]

Reviewer: ```suggestion
- the [Intel TDX] is required by the [ROFL TDX raw and container-based][rofl-types]
```

  apps.

To run SGX/TDX enclaves, your hardware must have SGX/TDX support, have it
enabled in your BIOS and have any additional drivers and software components

> properly installed and running.

Reviewer: Maybe also include that it must have all BIOS updates applied?

[attestation tool]: https://github.com/oasisprotocol/tools/tree/main/attestation-tool#readme
[binary]: https://github.com/oasisprotocol/tools/releases

> ## Trusted Domain Extension (TDX) {#tdx}

Reviewer: ```suggestion
## Trust Domain Extensions (TDX) {#tdx}
```

> Before proceeding to the TDX installation, **make sure you followed the SGX

Reviewer: ```suggestion
Before proceeding with the TDX installation, **make sure you followed the SGX
```

### BIOS configuration {#tdx-bios-configuration}

- **SGX**: ENABLE
- **SGX memory**: at least 256MB
- **SGX Auto MP Registration**: ENABLE
- **Hyper-Threading**: DISABLE
- **Intel SpeedStep**: DISABLE
- **SecureBoot**: DISABLE
- **All Internal Graphics**: DISABLE
- **Turbo Mode**: DISABLE
- **CPU AES**: ENABLE
- **TDX**: ENABLE
- **Memory Encryption (TME)**: ENABLE
- **Total Memory Encryption Bypass**: ENABLE
- **Total Memory Encryption Multi-Tenant (TME-MT)**: ENABLE
- **TME-MT memory integrity**: DISABLE
- **TDX Secure Arbitration Mode Loader (SEAM Loader)**: ENABLE
- **TME-MT/TDX key split**: non-zero value
- **ACPI S3 and deeper power states**: DISABLED

### Host OS setup

The following section contains summarized instructions for setting up an
environment for running ROFL node and other TDX services on Ubuntu 24.04 or
later. Check out the official [Canonical TDX repository] for details.

[Canonical TDX repository]: https://github.com/canonical/tdx

1. Add the following TDX PPAs to your APT sources and the keyring:
   
   ```shell
   sudo add-apt-repository ppa:kobuk-team/tdx-release
   sudo add-apt-repository ppa:kobuk-team/tdx-attestation-release
   sudo apt update
   ```

> 2. Install the trusted domain quote generation service and QEMU for running

Reviewer: ```suggestion
2. Install the TDX quote generation service and QEMU for running
```

>    guest containers:

Reviewer: ```suggestion
   guest virtual machines:
```

    ```shell
    sudo apt install tdx-qgs qemu-utils qemu-system-x86
    ```
   
3. Install a special TDX-enabled Linux kernel:

    ```shell
    sudo apt install linux-image-intel
    ```

3. Disable ACPI S3 (add kernel parameter: `nohibernate`):
 
   ```
   sed -i -E "s/GRUB_CMDLINE_LINUX=\"(.*)\"/GRUB_CMDLINE_LINUX=\"\1 nohibernate\"/g" /etc/default/grub
   update-grub
   ```

4. Make sure the non-root user running Oasis-node is a member of `sgx`,
   `sgx_prv` and `kvm` groups on host (access to `/dev/sgx*`, `/dev/kvm*` and
   `/dev/*vsock*` devices).

> 5. On Ubuntu 24.10 fix AppArmor profile for `bwrap`:

Reviewer: But this is also needed for SGX, should it be moved? Also this is already mentioned in the [System Configuration chapter](https://docs.oasis.io/node/run-your-node/prerequisites/system-configuration#apparmor-profiles), it should probably link there. Or where did you get this `bwrap-userns-restrict` file from?

Author: Moved under System configuration. It seems Ubuntu 24.10+ now already ships the user namespace restrict module for bwrap, you just need to enable it.

This is from @gw0's quick cheat sheet in https://github.com/oasisprotocol/docs/issues/1193#issuecomment-2804382269

   ```shell
   sudo ln -s /usr/share/apparmor/extra-profiles/bwrap-userns-restrict /etc/apparmor.d/
   ```

> 6. Reboot your system and select the new kernel.

Reviewer: Select the new kernel? It should be set as the default one.

Author: This wasn't the case on my machine :shrug: The new -intel kernels are just appended to the list and for the first time you need to select it. Then it's remembered.

Reviewer: No, I mean the guide should be updated to make sure it is set as default. The `canonical/tdx` repository includes a script which changes the default kernel in grub.

Author: Extracted the required snippet to detect it and set it.

### Check TDX Setup

To check whether your system is ready to launch TDX containers, you should see
an output like this:

```shell
sudo dmesg | grep -i tdx
```

```
[    1.368184] virt/tdx: BIOS enabled: private KeyID range [32, 64)
[    1.368188] virt/tdx: Disable ACPI S3. Turn off TDX in the BIOS to use ACPI S3.
[    9.483666] virt/tdx: TDX module: attributes 0x0, vendor_id 0x8086, major_version 1, minor_version 5, build_date 20240725, build_num 784
[    9.483672] virt/tdx: CMR: [0x100000, 0x77800000)
[    9.483674] virt/tdx: CMR: [0x100000000, 0x3ffe000000)
[   10.587167] virt/tdx: 1042424 KB allocated for PAMT
[   10.587173] virt/tdx: module initialized
[ 1747.370008]  tdx_handle_ept_violation+0xe7/0x1b0 [kvm_intel]

> [ 1747.370029]  tdx_handle_exit+0x1cf/0x240 [kvm_intel]

Reviewer: Remove these last two, this seems like a stack trace from some kernel errror/debug message.
